src/app-layer-parser.c

```diff
@@ -935,6 +935,11 @@ void AppLayerParserTransactionsCleanup(Flow *f, const uint8_t pkt_dir)
     SCLogDebug("start min %"PRIu64, min);
     bool skipped = false;
     // const bool support_files = AppLayerParserSupportsFiles(f->proto, f->alproto);
+    size_t tofree_nb = 0;
+    uint64_t *tofree = SCMalloc(total_txs * sizeof(uint64_t));
```

Comment:
explanation for this whole logic is missing, but if we need something like it I don't want to see this allocation, but instead have something we can reuse and grow on demand

Reply:
Putting here some extensive explanation, to see what is the right solution.

App-layer protocols have transaction iterators.
These transaction iterators are implemented by the app-layer, based on the structure used for the list of transactions : it may be a C linked list with pointers, a rust VecDeque, etc...
These tx iterators are used in different places: detection, logging, ... and cleanup.
Cleanup removes the transactions from their list structure.

So, we face the classical problem coming of removing elements of a list while iterating over it.

The current solution, for rust, implemented by a trait (yeah 2 levels of abstraction), is to propose the "next" element (variable state) as the same that was returned.
So, in most cases, next call of the iterator, restarts from the previously returned element, and makes a dummy iteration of the loop in `get_transaction_iterator` to move forward
In the case where the tx got removed, there is no dummy iteration and it returns quickly.

I want to optimize the first case which happens more, and remain correct...

I propose the solution to not have removal of transactions while iterating over their list, as it is a classic treacherous pattern.
(An alternative fix could be to change `get_transaction_iterator` so that it checks if a tx got removed between 2 calls)
So, the solution is to have a first loop iterating over the transactions and retaining the tx ids to remove.
And then iterate over these tx ids to remove the transactions.

Why do you prefer a long-lived allocation over a shorter one ?